File: lc_go/lcGo/dp/maxAbsoluteSum.go
package main

func maxAbsoluteSum(nums []int) int {
	miEn, miSum, mxSum, mxEn := 0, 0, 0, 0

	for _, x := range nums {
		mxEn = max(x, x+mxEn)
		mxSum = max(mxEn, mxSum)
		miEn = min(x, x+miEn)
		miSum = min(miSum, miEn)
	}
	return max(mxSum, -miSum)
}
func main() {
	nums := []int{2, -5, 1, -4, 3, -2}
	println(maxAbsoluteSum(nums))
}

File: lc_cpp/lcCpp/dp/maxAbsoluteSum.cpp
//
// Created by 24276 on 2025/8/12.
//
#include <vector>

using namespace std;

class max_absolute_sum {
public:
    int maxAbsoluteSum(vector<int>& nums) {
        int mx_sum = 0, min_sum = 0;
        int mx_end = 0, min_end = 0;
        for(int v : nums ) {
            mx_end = max(v, mx_end + v);
            mx_sum = max(mx_sum, mx_end);
            min_end = min(v, min_end + v);
            min_sum = min(min_sum, min_end);
        }
        return max(mx_sum, -min_sum);
    }
};

int main() {
    max_absolute_sum ps;
    vector<int> nums = {1, -3, 2, 3, -4};
    vector <int> nums2 = {2,-5,1,-4,3,-2};
    ps.maxAbsoluteSum(nums2);
    return 0;
}
